### PR TITLE
HDDS-9146. Potential data loss with HSync due to deletedTable entry having the same block as keyTable entry's

### PR DESCRIPTION
**Pull request by smengcl**
## What changes were proposed in this pull request?

See JIRA description for the background and a repro on this one.

- [x] Filter out blocks in the old key that would be put into `deletedTable` in `OMKeyCommitRequest` and `OMKeyCommitRequestWithFSO`.
- [x] Add debug logging in the case of any type of key commit, successful or not.
- [x] Improve and implement `toString()` for various classes to facilitate logging.
- [x] Expose `awaitDoubleBufferFlush()` method in `OzoneManager` that works with both Ratis and non-Ratis. Currently used by the integration test addition to eliminate flakiness due to the timing of double buffer flush.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9146

## How was this patch tested?

- [x] Added unit test case in `TestOMKeyCommitRequest`.
- [x] New integration test case added in `TestHSync` (similar to the repro, checks `deletedTable`).

### CODE REVIEW

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestHSync.java [hunk 1]
```diff
@@ -131,6 +146,47 @@ public static void teardown() {
...
+
+    OMMetadataManager ommm = cluster.getOzoneManager().getMetadataManager();
+    // deletedTable should not have an entry for file at all in this case
+    try (TableIterator<String,
+        ? extends Table.KeyValue<String, RepeatedOmKeyInfo>>
+        tableIter = ommm.getDeletedTable().iterator()) {
+      while (tableIter.hasNext()) {
+        Table.KeyValue<String, RepeatedOmKeyInfo> kv = tableIter.next();
+        String key = kv.getKey();
+        if (key.startsWith(file.toString())) {
```
Comment (jojochuang) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestHSync.java:179]: shouldn't it fail if deletedTable is not empty? Also key string will not match file string anyway.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java [hunk 2]
```diff
@@ -301,6 +323,12 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
               exception, getOmRequest().getUserInfo()));
       processResult(commitKeyRequest, volumeName, bucketName, keyName,
           omMetrics, exception, omKeyInfo, result);
+    } else {
+      // Debug logging for HSync
+      if (LOG.isDebugEnabled()) {
```
Comment (jojochuang) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:328]: this check is redundant.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyLocationInfoGroup.java [hunk 3]
```diff
@@ -177,15 +184,11 @@ void addAll(long versionToAdd, List<OmKeyLocationInfo> locationInfoList) {
...
-    StringBuilder sb = new StringBuilder();
-    sb.append("version:").append(version).append(" ");
-    sb.append("isMultipartKey:").append(isMultipartKey).append(" ");
-    for (List<OmKeyLocationInfo> kliList : locationVersionMap.values()) {
-      for (OmKeyLocationInfo kli: kliList) {
-        sb.append(kli.getLocalID()).append(" || ");
-      }
-    }
-    return sb.toString();
+    return "OmKeyLocationInfoGroup{" +
```
Comment (jojochuang) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyLocationInfoGroup.java:187]: why the change? StringBuilder looks more elegant and performant.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestHSync.java (hunk 1 above)
Comment (smengcl) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestHSync.java:179]: I can't ensure if any other tests in the same test class doesn't trigger actual block deletion. Thus I am not enforcing emptiness on the deletedTable

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java (hunk 2 above)
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:328]: I'm fine either way. I was following the same exercise applied elsewhere, mostly to avoid parameter evaluation I assume:

https://github.com/apache/ozone/blob/a66e43e2838f087229cb4b0fac1e1dc969ab80dd/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/BackgroundService.java#L103-L105

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestHSync.java (hunk 1 above)
Comment (smengcl) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestHSync.java:179]: > Also key string will not match file string anyway

It works. Try commenting out this line, run the test and see it fail:

https://github.com/apache/ozone/blob/edad661120a6b7e4490754323ed241423447a96e/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestWithFSO.java#L216

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java (hunk 2 above)
Comment (jojochuang) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:328]: There's no parameter evaluation cost if not in debug log level. https://www.slf4j.org/faq.html#logging_performance

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyLocationInfoGroup.java (hunk 3 above)
Comment (smengcl) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyLocationInfoGroup.java:187]: It was lacking detailed `locationVersionMap` output (no container ID, no bcsId). Also not formatted well so I am just using IntelliJ auto-generated `toString()` to replace it.

AFAIK using `StringBuilder` outside of a big loop should not make too much of a performance difference nowadays.

I might just restore it but add those interested fields.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java (hunk 2 above)
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:328]: Yup.. I also wrote it this way in anticipation of other reviewers bugging me about adding this extra `if` check (been asked to do that before).

I will just remove it in this case.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyLocationInfoGroup.java (hunk 3 above)
Comment (smengcl) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyLocationInfoGroup.java]: Done in 55b932fcd0432afc70e80f6a3b5868a276f5d0a0

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java (hunk 2 above)
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java]: Done in d058d50b7f5dc07ebf26437ec5b14cb6e0f95a9b

**Comment by smengcl** [comment]
Thanks @jojochuang for the review. I've addressed the 3 comments you have so far.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyInfo.java [hunk 4]
```diff
@@ -425,6 +425,25 @@ public FileChecksum getFileChecksum() {
...
+    return "OmKeyInfo{" +
+        "volumeName='" + volumeName + '\'' +
+        ", bucketName='" + bucketName + '\'' +
+        ", keyName='" + keyName + '\'' +
+        ", dataSize=" + dataSize +
+        ", keyLocationVersions=" + keyLocationVersions +
+        ", creationTime=" + creationTime +
+        ", modificationTime=" + modificationTime +
+        ", replicationConfig=" + replicationConfig +
+        ", encInfo=" + encInfo +
```
Comment (sumitagrawl) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyInfo.java:439]: Encryption info seems sensitive to be logged, please check and remove this.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java [hunk 5]
```diff
@@ -242,7 +253,16 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
         if (null == oldKeyVersionsToDeleteMap) {
           oldKeyVersionsToDeleteMap = new HashMap<>();
         }
-        oldKeyVersionsToDeleteMap.put(delKeyName, oldVerKeyInfo);
+
+        // Remove any block from oldVerKeyInfo that share the same container ID
+        // and local ID with omKeyInfo blocks'.
```
Comment (sumitagrawl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:258]: The fix is good to avoid deletion of blocks in generic. I think actual root cause can also be fixed. Attached fix for this.

[hsync_delete_issue.patch](https://github.com/apache/ozone/files/12312760/hsync_delete_issue.patch)

**Review by sumitagrawl — COMMENTED** [review]
@smengcl Thanks for working over this, have few comment.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyInfo.java (hunk 4 above)
Comment (smengcl) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyInfo.java:439]: Good catch. I have changed it into logging `null` or `<REDACTED>`:

```
", encInfo=" + (encInfo == null ? "null" : "<REDACTED>") +
```

Comment (smengcl) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyInfo.java:439]: Done in 3be3d149431e04309cdd341685df51605ce5a8b0

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java (hunk 5 above)
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:258]: Thanks @sumitagrawl . Similar idea also occurred to us but I wasn't proficient in HSync block commit flow so @errose28 , @jojochuang and I decided to take the generic filtering approach, for now.

1. I briefly tried out your patch on top of this PR branch. I get the idea. However, if you comment out the call to the filtering logic:

https://github.com/apache/ozone/blob/5661ad98616245606baffd09c86472d6af9c2968/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestWithFSO.java#L214

and run the new integration test, you'll see `OMKeyCommitRequestWithFSO` is still putting the key block into `deletedTable`:

```
org.opentest4j.AssertionFailedError: deletedTable should not have such entry. key = /volume13960/bucket38943/file-hsync-then-close/4611686018427388673

	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:39)
	at org.junit.jupiter.api.Assertions.fail(Assertions.java:134)
	at org.apache.hadoop.fs.ozone.TestHSync.testKeyHSyncThenClose(TestHSync.java:184)
```

That is because when it is reading the `omKeyInfo` during the second non-HSync commit, `omKeyInfo.getMetadata()` is empty. So `isPreviousCommitHsync` is always `false` (at least in FSO).

In turn this is because `omKeyInfo` is retrieved from `OpenKeyTable`:

https://github.com/apache/ozone/blob/5661ad98616245606baffd09c86472d6af9c2968/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestWithFSO.java#L153-L154

and in KeyCommitRequest we only put the metadata in `KeyTable`, not `OpenKeyTable`:

https://github.com/apache/ozone/blob/5661ad98616245606baffd09c86472d6af9c2968/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestWithFSO.java#L252-L253

Therefore, the patch itself it looks incomplete. Would you raise a another PR once this filtering patch is merged?

2. Another problem is, currently I see no check to ensure the same `clientId` is doing the commit in HSync key commit flow. Therefore, if another client with different ClientID (rogue or not) tries to commit this key without HSync, `isPreviousCommitHsync` will be `false` according to your patch. Then again it will fall into the second branch, which queues the block for deletion and in turn cause data loss if the block filtering logic does not exist.

Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:258]: All in all, I think we will need to further:

1) Update OpenKeyTable with the HSync metadata as well? (and check if such action would lead to further unintended side effects)
2) Improve or fix the clientId check
3) Sort out the HSync and non-HSync key commit flow to make sure no other code paths would lead to unintended consequences like this. And fix them if we found any other issues.

in a series of follow-up patches.

Comment (jojochuang) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:258]: isPreviousCommitHsync needs to be true whenever HSYNC_CLIENT_ID has a value.

**Comment by smengcl** [comment]
Merged HDDS-9148 to the PR branch, pls take another look @jojochuang @sumitagrawl .

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java (hunk 5 above)
Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:269]: Hsync flow: openKeyTable/openFileTable will have hsync flag whenever first commit comes with hsync feature
1. openKeyTable: when first commit comes with hsync flag true (as hsync feature), it will add to metadata of open key table
2. Multiple commits can come with hsync flag true (always)
3. When final commit comes (with hsync flag false), it will retrive same flag from metadata of open key table

Related to ClientId,
- commit always retrieve omKeyInfo from with clientId as full keyName, else commit will fail. So there can not be any other annonymous clientId
`String dbOpenKey = omMetadataManager.getOpenKey(volumeName, bucketName,
          keyName, commitKeyRequest.getClientID());`

We can take up this as separate verifying if any other issue, and this PR looks good to me.

**Review by sumitagrawl — APPROVED** [review]
@smengcl LGTM +1

**Comment by smengcl** [comment]
Thanks @sumitagrawl @jojochuang for the reviews.

@sumitagrawl Pls feel free to take over further fixes and improvements around hsync from here.
